libsolidity/analysis/NameAndTypeResolver.cpp

```diff
@@ -703,6 +703,26 @@ void DeclarationRegistrationHelper::registerDeclaration(Declaration& _declaratio
 	solAssert(m_currentScope && m_scopes.count(m_currentScope), "No current scope.");
 	solAssert(m_currentScope == _declaration.scope(), "Unexpected current scope.");
 
+	// Functions defined inside quantifiers are available in the scope containing the quantifier
+	auto const* quantifier = dynamic_cast<ForAllQuantifier const*>(m_currentScope);
+	auto const* functionDefinition = dynamic_cast<FunctionDefinition const*>(&_declaration);
+	if (quantifier && functionDefinition)
+	{
+		solAssert(quantifier->scope());
+		solAssert(
+			// forall quantifiers cannot be used in block scope so the declaration is always active.
+			!dynamic_cast<Block const*>(quantifier->scope()) &&
+			!dynamic_cast<ForStatement const*>(quantifier->scope())
+		);
+		registerDeclaration(*m_scopes.at(quantifier->scope()), _declaration, nullptr, nullptr, false /* inactive */, m_errorReporter);
+
+		// TMP: Should I adjust the scope to match visibility? Here or in the Scoper?
+		//_declaration.annotation().scope = quantifier->scope();
```

Comment:
Since this is experimental-only, we can also wait and see if and when this will cause problems.
In general rather nice that this approach doesn't seem to take *that* much hacking in the name and type resolver and all is localized here!

Reply:
> Since this is experimental-only, we can also wait and see if and when this will cause problems.

After taking a closer look at how the registration and `scope()` annotation work, I think this will not cause any problems now, but it may make the interpretation of `scope()` on functions a bit error prone.

The declaration->scope mapping from annotations seems to be the used by name resolver only to construct the reverse scope->delacration mapping and then it's that reverse mapping that's used for name lookups. So this will work just fine.

The issue with code accessing `scope()` directly, on the other hand, is that it will have to explicitly account for the quantifier. Right now I allowed the quantifier only on free functions and I have not seen any checks that would be affected, but when we reintroduce contracts, this will break a lot of checks whether a function belongs to a contract. Currently they assume that the contract must be the function's scope.

I made a quick attempt to change quantified function scope in `Scoper`, but this breaks `DeclarationRegistrationHelper`'s assumptions about tree traversal - it assumes that scopes follow the nesting of `ScopeOpener`s. This can be adjusted, but I did not want the special-casing to spread to multiple places so I reverted that change in the end.

One final thing to keep in mind is also that the scope is visible in the AST output. So it may also break external tools that make assumptions about how functions are nested.